src/corefx/System.Globalization.Native/collation.cpp

```diff
@@ -0,0 +1,254 @@
+
+    if (U_SUCCESS(err))
+    {
+        UStringSearch* pSearch = usearch_openFromCollator(lpTarget, -1, lpSource, cwSourceLength, pColl, nullptr, &err);
+        int32_t idx = USEARCH_DONE;
+
+        if (U_SUCCESS(err))
+        {
+            idx = usearch_first(pSearch, &err);
+
+            if (idx == 0)
+            {
+                result = TRUE;
+            }
+            else
+            {
+                UCollationElements* pCollElem = ucol_openElements(pColl, lpSource, idx, &err);
+
+                if (U_SUCCESS(err))
+                {
+                    int32_t curCollElem = UCOL_NULLORDER;
+
+                    result = TRUE;
+
+                    while((curCollElem = ucol_next(pCollElem, &err)) != UCOL_NULLORDER)
+                    {
+                        if (curCollElem != 0 || U_FAILURE(err))
+                        {
+                            // Non ignorable collation element found between start of the string and the first match for lpTarget.
+                            result = FALSE;
+                            break;
+                        }
+                    }
+
+                    ucol_closeElements(pCollElem);
+                }
+            }
+
+            usearch_close(pSearch);
+        }
+
+        ucol_close(pColl);
+    }
+
+    return result;
+}
+
+/*
+ Return value is a "Win32 BOOL" (1 = true, 0 = false)
+ */
+extern "C" int32_t EndsWith(const char* lpLocaleName, const UChar* lpTarget, const UChar* lpSource, int32_t cwSourceLength, int32_t options)
+{
+    int32_t result = FALSE;
+    UErrorCode err = U_ZERO_ERROR;
+    UCollator* pColl = GetCollatorForLocaleAndOptions(lpLocaleName, options, &err);
+
+    if (U_SUCCESS(err))
+    {
+        UStringSearch* pSearch = usearch_openFromCollator(lpTarget, -1, lpSource, cwSourceLength, pColl, nullptr, &err);
+        int32_t idx = USEARCH_DONE;
+
+        if (U_SUCCESS(err))
+        {
+            idx = usearch_last(pSearch, &err);
+
+            if (idx != USEARCH_DONE)
+            {
+                if ((idx + usearch_getMatchedLength(pSearch)) == cwSourceLength)
+                {
+                    result = TRUE;
+                }
+
+                // TODO (dotnet/corefx#3467): We should do something similar to what StartsWith does where we can ignore
+                // some collation elements at the end of te string if they are zero.
+            }
+
+            usearch_close(pSearch);
+        }
+
+        ucol_close(pColl);
+    }
+
+    return result;
+}
+
+extern "C" int32_t GetSortKey(const char* lpLocaleName, const UChar* lpStr, int32_t cwStrLength, uint8_t* sortKey, int32_t cbSortKeyLength, int32_t options)
+{
+    UErrorCode err = U_ZERO_ERROR;
+    UCollator* pColl = GetCollatorForLocaleAndOptions(lpLocaleName, options, &err);
+    int32_t result = 0;
+
+    if (U_SUCCESS(err))
+    {
+        result = ucol_getSortKey(pColl, lpStr, cwStrLength, sortKey, cbSortKeyLength);
+    }
```

Comment:
Not at all, just wanted to check.  Wasn't sure if maybe we had to keep it open for the result to still be valid or something.